### ISSUE TITLE
Server does not react SIGTERM

### ISSUE THREAD

**Issue by ava57r** [qdrant/qdrant#2075]
qdrant version 1.2.2

## Current Behavior
2 servers cluster does not shutdown after `Ctrl+C`

## Steps to Reproduce
1. Load big collection
2.  don't touch some days.
3. press Ctrl+C  on second server (not leader)
4. press Ctrl+C on first server (leader)

second server repeats `starting a new election` again and again

```log
2023-06-13T13:43:19.006Z INFO  raft::raft] starting a new election, raft_id: ..
[2023-06-13T13:43:19.006Z INFO  raft::raft] became candidate at term 460, raft_id: ...
[2023-06-13T13:43:19.006Z INFO  raft::raft] broadcasting vote request, raft_id: ..., term: 460, type: MsgRequestVote
[2023-06-13T13:43:19.006Z DEBUG qdrant::consensus] Changing hard state. New hard state: HardState { term: 460, vote: .., commit: 9 }
[2023-06-13T13:43:19.007Z DEBUG hyper::client::connect::dns] resolving host="host"
[2023-06-13T13:43:19.008Z DEBUG hyper::client::connect::http] connecting to host
[2023-06-13T13:43:22.446Z INFO  raft::raft] starting a new election, raft_id: ...
[2023-06-13T13:43:22.446Z INFO  raft::raft] became candidate at term 461, raft_id: ...
[2023-06-13T13:43:22.446Z INFO  raft::raft] broadcasting vote request, raft_id: ..., term: 461, type: MsgRequestVote
[2023-06-13T13:43:22.446Z DEBUG qdrant::consensus] Changing hard state. New hard state: HardState { term: 461, vote: ..., commit: 9 }
```

first server 

```log
2023-06-13T13:43:21.814Z DEBUG hyper::client::connect::dns] resolving host="host"
[2023-06-13T13:43:21.817Z DEBUG hyper::client::connect::http] connecting to IP
```

I stopped it on first server after `SIGKILL`

**Comment by Livingstone110** [comment]
> qdrant version 1.2.2
> 
> 
> 
> ## Current Behavior
> 
> 2 servers cluster does not shutdown after `Ctrl+C`
> 
> 
> 
> ## Steps to Reproduce
> 
> 1. Load big collection
> 
> 2.  don't touch some days.
> 
> 3. press Ctrl+C  on second server (not leader)
> 
> 4. press Ctrl+C on first server (leader)
> 
> 
> 
> second server repeats `starting a new election` again and again
> 
> 
> 
> ```log
> 
> 2023-06-13T13:43:19.006Z INFO  raft::raft] starting a new election, raft_id: ..
> 
> [2023-06-13T13:43:19.006Z INFO  raft::raft] became candidate at term 460, raft_id: ...
> 
> [2023-06-13T13:43:19.006Z INFO  raft::raft] broadcasting vote request, raft_id: ..., term: 460, type: MsgRequestVote
> 
> [2023-06-13T13:43:19.006Z DEBUG qdrant::consensus] Changing hard state. New hard state: HardState { term: 460, vote: .., commit: 9 }
> 
> [2023-06-13T13:43:19.007Z DEBUG hyper::client::connect::dns] resolving host="host"
> 
> [2023-06-13T13:43:19.008Z DEBUG hyper::client::connect::http] connecting to host
> 
> [2023-06-13T13:43:22.446Z INFO  raft::raft] starting a new election, raft_id: ...
> 
> [2023-06-13T13:43:22.446Z INFO  raft::raft] became candidate at term 461, raft_id: ...
> 
> [2023-06-13T13:43:22.446Z INFO  raft::raft] broadcasting vote request, raft_id: ..., term: 461, type: MsgRequestVote
> 
> [2023-06-13T13:43:22.446Z DEBUG qdrant::consensus] Changing hard state. New hard state: HardState { term: 461, vote: ..., commit: 9 }
> 
> ```
> 
> 
> 
> first server 
> 
> 
> 
> ```log
> 
> 2023-06-13T13:43:21.814Z DEBUG hyper::client::connect::dns] resolving host="host"
> 
> [2023-06-13T13:43:21.817Z DEBUG hyper::client::connect::http] connecting to IP
> 
> ```
> 
> 
> 
> I stopped it on first server after `SIGKILL`

**Comment by generall** [comment]
We have support for `SIGINT`, if it helps

**Comment by ava57r** [comment]
if it runs with `systemd` that does not stop correctly.
I changed default `KillSignal = SIGINT`

**Comment by ava57r** [comment]
> We have support for `SIGINT`, if it helps

Could We add more handlers for signals?

**Comment by bazhenov** [comment]
> Could We add more handlers for signals?

I don't think this will help. Based on my limited knowledge, symptoms show that main thread is waiting on one or more threads to join. `starting a new election` is probably from the consensus loop which is not joined in the main at the moment.

Could please check if `kill -s INT [PID]` is working as expected or not?

**Comment by bazhenov** [comment]
Also if you are on macOS it's useful to get output of 
```console
$ sudo sample [PID] 0
```
to get stacktraces of all the running threads

**Comment by ava57r** [comment]
> > Could We add more handlers for signals?
> 
> I don't think this will help. Based on my limited knowledge, symptoms show that main thread is waiting on one or more threads to join. `starting a new election` is probably from the consensus loop which is not joined in the main at the moment, so I bet it has nothing to do with this problem.
> 
> Could please check if `kill -s INT [PID]` is working as expected or not?

I checked. `SIGINT` works.
I think "more signals" is for another issue.

**Comment by bazhenov** [comment]
> I think "more signals" is for another issue.

Probably 😀. Still, there is some problem with handling `SIGINT`. I suppose <kbd>Ctrl</kbd>+<kbd>C</kbd> behavior must be consistent with `kill -s INT ...`

**Comment by bazhenov** [comment]
@ava57r what OS is in question? It would be great if you could provide stack traces of the threads. It will help tremendously with diagnosing the issue.

**Comment by ava57r** [comment]
It is Linux

**Comment by bazhenov** [comment]
On a Linux machine you can create core dump and then use `gdb` to print backtraces of all the threads:

```console
$ gcore [PID]
$ gdb -c core-file
```

`gdb` has commands `info threads`, `thread [ID]`, and `bt` which allows to list all threads, select thread and print backtrace of a thread respectively.
